File: Folicon_Native/Modules/TMDB.vb

Imports System.Collections.ObjectModel
Imports System.Net.TMDb
Imports System.Threading

Namespace Modules

    Module TMDB
        Dim posterBase As String = "https://image.tmdb.org/t/p/w200"
        Public Class ResultResponse
            Public Property Result As Object
            Public Property MediaType As String
        End Class

        Public Async Function SearchIt(query As String, serviceClient As ServiceClient) As Task(Of ResultResponse)
            Dim r = Nothing
            Dim mt As String = ""
            If SearchMod = "Movie" Then

                If query.ToLower.Contains("collection") Then
                    r = Await serviceClient.Collections.SearchAsync(query, "en-US", 1, CancellationToken.None)
                    mt = "Collection"
                Else
                    r =
                        Await _
                            serviceClient.Movies.SearchAsync(query, "en-US", False, Nothing, True, 1, CancellationToken.None)
                    mt = "Movie"
                End If
            ElseIf SearchMod = "TV" Then
                r = Await serviceClient.Shows.SearchAsync(query, "en-US", Nothing, True, 1, CancellationToken.None)
                mt = "TV"
            ElseIf SearchMod = "Auto (Movies & TV Shows)" Then
                ' TODO: Write logic to search for all
                Dim iop As Net.TMDb.Resources
                r = Await serviceClient.SearchAsync(query, "en-US", False, 1, CancellationToken.None)
                mt = "Auto (Movies & TV Shows)"
            End If
            Searchresultob = r
            Dim response As New ResultResponse() With {.Result = r, .MediaType = mt}
            Return response
        End Function


        Public Sub ResultPicked(result As Object, resultType As String, pickIndex As Integer)
            If result.Results(pickIndex).Poster Is Nothing Then
                Throw New Exception("NoPoster")
            End If
            Dim localPosterPath = SelectedFolderPath & "\" & Fnames(FolderNameIndex) & "\" & Fnames(FolderNameIndex) &
                                  ".png"
            Dim folderPath = SelectedFolderPath & "\" & Fnames(FolderNameIndex)
            Dim folderName = Fnames(FolderNameIndex)
            Dim posterUrl = String.Concat("http://image.tmdb.org/t/p/original", result.Results(pickIndex).Poster)
            'If Not String.IsNullOrEmpty(posterUrl)
            If resultType = "TV" Then
                Dim searchResult = DirectCast(result, Shows)
                Dim pickedResult = searchResult.Results(pickIndex)
                AddToPickedListDataTable(localPosterPath,
                                         pickedResult.Name, pickedResult.VoteAverage,
                                         folderPath,
                                         folderName, pickedResult.FirstAirDate.Value.Year)
            ElseIf resultType = "Movie" Then
                Dim searchResult = DirectCast(result, Movies)
                Dim pickedResult = searchResult.Results(pickIndex)

                AddToPickedListDataTable(localPosterPath,
                                         pickedResult.Title, pickedResult.VoteAverage,
                                         folderPath, folderName, pickedResult.ReleaseDate.Value.Year)
            ElseIf resultType = "Collection" Then

                Dim searchResult = DirectCast(result, Collections)
                Dim pickedResult = searchResult.Results(pickIndex)
                AddToPickedListDataTable(localPosterPath,
                                         pickedResult.Name, "",
                                         folderPath, folderName)
            ElseIf resultType = "Auto (Movies & TV Shows)" Then
                Dim mediaType = result.Results(pickIndex).GetType()
                If mediaType.Name = "Show" Then
                    Dim pickedResult = DirectCast(result.Results(pickIndex), Show)
                    AddToPickedListDataTable(localPosterPath,
                                             pickedResult.Name, pickedResult.VoteAverage,
                                             folderPath,
                                             folderName, pickedResult.FirstAirDate.Value.Year)
                ElseIf mediaType.Name = "Movie" Then
                    Dim pickedResult = DirectCast(result.Results(pickIndex), Movie)
                    AddToPickedListDataTable(localPosterPath,
                                             pickedResult.Title, pickedResult.VoteAverage,
                                             folderPath, folderName, pickedResult.ReleaseDate.Value.Year)
                End If
            End If
            FolderProcessedCount += 1
            Dim tempImage As New ImageToDownload() With {
                    .LocalPath = localPosterPath,
                    .RemotePath = posterUrl}
            ImgDownloadList.Add(tempImage)
            'End If
        End Sub

        Public Function ExtractTvDetailsIntoListItem(result As Shows) As ObservableCollection(Of ListItem)
            Dim items As New ObservableCollection(Of ListItem)()
            Dim mediaName As String
            Dim year As String
            Dim rating As String
            Dim overview As String
            Dim poster As String
            For Each item In result.Results
                mediaName = item.Name
                year = If(item.FirstAirDate IsNot Nothing, item.FirstAirDate.Value.Year, "")
                rating = item.VoteAverage
                overview = item.Overview
                poster = If(item.Poster IsNot Nothing, posterBase & item.Poster, Nothing)
                items.Add(New ListItem(mediaName, year, rating, overview, poster))
            Next
            Return items
        End Function

        Public Function ExtractMoviesDetailsIntoListItem(result As Movies) As ObservableCollection(Of ListItem)
            Dim items As New ObservableCollection(Of ListItem)()
            Dim mediaName As String
            Dim year As String
            Dim rating As String
            Dim overview As String
            Dim poster As String
            For Each item In result.Results
                mediaName = item.Title
                year = If(item.ReleaseDate IsNot Nothing, item.ReleaseDate.Value.Year, "")
                rating = item.VoteAverage
                overview = item.Overview
                poster = If(item.Poster IsNot Nothing, posterBase & item.Poster, Nothing)
                items.Add(New ListItem(mediaName, year, rating, overview, poster))
            Next
            Return items
        End Function

        Public Function ExtractResourceDetailsIntoListItem(result As Resources) As ObservableCollection(Of ListItem)
            Dim items As New ObservableCollection(Of ListItem)()
            Dim mediaName = ""
            Dim year = ""
            Dim rating = ""
            Dim overview = ""
            Dim poster = Nothing
            For Each item In result.Results
                Dim mediaType = item.GetType()
                If mediaType.Name = "Show" Then
                    Dim res As Show = item
                    mediaName = res.Name
                    year = If(res.FirstAirDate IsNot Nothing, res.FirstAirDate.Value.Year, "")
                    rating = res.VoteAverage
                    overview = res.Overview
                    poster = If(res.Poster IsNot Nothing, posterBase & res.Poster, Nothing)
                ElseIf mediaType.Name = "Movie" Then
                    Dim res As Movie = item
                    mediaName = res.Title
                    year = If(res.ReleaseDate IsNot Nothing, res.ReleaseDate.Value.Year, "")
                    rating = res.VoteAverage
                    overview = res.Overview
                    poster = If(res.Poster IsNot Nothing, posterBase & res.Poster, Nothing)
                End If

                items.Add(New ListItem(mediaName, year, rating, overview, poster))
            Next
            Return items
        End Function

        Public Function ExtractCollectionDetailsIntoListItem(result As Collections) As ObservableCollection(Of ListItem)
            Dim items As New ObservableCollection(Of ListItem)()
            Dim mediaName As String
            Dim year As String
            Dim rating As String
            Dim overview As String
            Dim poster As String
            For Each item In result.Results
                mediaName = item.Name
                year = ""
                rating = ""
                overview = ""
                poster = If(item.Poster IsNot Nothing, posterBase & item.Poster, Nothing)
                items.Add(New ListItem(mediaName, year, rating, overview, poster))
            Next
            Return items
        End Function
    End Module
End NameSpace

File: FoliCon/Modules/TMDB.cs
using FoliCon.Modules.utils;
using NLog;
using Collection = TMDbLib.Objects.Collections.Collection;
using Logger = NLog.Logger;

namespace FoliCon.Modules;

public class Tmdb
{
    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
    private const string PosterBase = "http://image.tmdb.org/t/p/original";
    private const string SmallPosterBase = "https://image.tmdb.org/t/p/w200";
    private readonly List<ImageToDownload> _imgDownloadList;
    private readonly DataTable _listDataTable;
    private readonly TMDbClient _serviceClient;

    public TMDbClient GetClient()
    {
        return _serviceClient;
    }

    /// <summary>
    /// TMDB Helper Class for Working with IGDB API efficiently for this project.
    /// </summary>
    /// <param name="serviceClient">Initialized TMDB Client</param>
    /// <param name="listDataTable">DataTable that stores all the Picked Results.</param>
    /// <param name="imgDownloadList">List that stores all the images to download.</param>
    public Tmdb(ref TMDbClient serviceClient, ref DataTable listDataTable,
        ref List<ImageToDownload> imgDownloadList)
    {
        Logger.Debug("Initializing TMDB Helper Class");
        _listDataTable = listDataTable;
        _serviceClient = serviceClient;
        _imgDownloadList = imgDownloadList;
        _ = _serviceClient.GetConfigAsync().Result;
        Logger.Debug("Initialized TMDB Helper Class");
    }
    public ImagesWithId SearchMovieImages(int tvId)
    {
        return _serviceClient.GetMovieImagesAsync(tvId).Result;
    }
    public ImagesWithId SearchTvImages(int movieId)
    {
        return _serviceClient.GetTvShowImagesAsync(movieId).Result;
    }
    public ImagesWithId SearchCollectionImages(int collectionId)
    {
        return _serviceClient.GetCollectionImagesAsync(collectionId).Result;
    }

        public static ObservableCollection<ListItem> ExtractCollectionDetailsIntoListItem(
            SearchContainer<SearchCollection> result)
        {
            var items = new ObservableCollection<ListItem>();
            foreach (var item in result.Results)
            {
                var mediaName = item.Name;
                const string year = "";
                const string rating = "";
                const string overview = "";
                var poster = Convert.ToString(item.PosterPath != null ? SmallPosterBase + item.PosterPath : null, CultureInfo.InvariantCulture);
                items.Add(new ListItem(mediaName, year, rating, overview, poster, "", item.Id.ToString()));
            }

        return items;
    }

        public static ObservableCollection<ListItem> ExtractMoviesDetailsIntoListItem(
            SearchContainer<SearchMovie> result)
        {
            var items = new ObservableCollection<ListItem>();
            foreach (var item in result.Results)
            {
                var mediaName = item.Title;
                var year = item.ReleaseDate != null ? item.ReleaseDate.Value.Year.ToString(CultureInfo.InvariantCulture) : "";
                var rating = item.VoteAverage.ToString(CultureInfo.CurrentCulture);
                var overview = item.Overview;
                var poster = item.PosterPath != null ? SmallPosterBase + item.PosterPath : null;
                items.Add(new ListItem(mediaName, year, rating, overview, poster, "", item.Id.ToString()));
            }

        return items;
    }

        public static ObservableCollection<ListItem> ExtractResourceDetailsIntoListItem(
            SearchContainer<SearchBase> result)
        {
            Logger.Debug("Extracting Resource Details into List Item");
            var items = new ObservableCollection<ListItem>();
            var mediaName = "";
            var year = "";
            var rating = "";
            var overview = "";
            string poster = null;
            int id = 0;
            foreach (var item in result.Results)
            {
                var mediaType = item.MediaType;
                
                Logger.Debug("Media Type: {MediaType}", mediaType);
                switch (mediaType)
                {
                    case MediaType.Tv:
                        {
                            var res = (SearchTv)item;
                            id = res.Id;
                            mediaName = res.Name;
                            year = res.FirstAirDate != null ? res.FirstAirDate.Value.Year.ToString(CultureInfo.InvariantCulture) : "";
                            rating = res.VoteAverage.ToString(CultureInfo.CurrentCulture);
                            overview = res.Overview;
                            poster = res.PosterPath != null ? SmallPosterBase + res.PosterPath : null;
                            break;
                        }
                    case MediaType.Movie:
                        {
                            var res = (SearchMovie)item;
                            id = res.Id;
                            mediaName = res.Title;
                            year = res.ReleaseDate != null ? res.ReleaseDate.Value.Year.ToString(CultureInfo.InvariantCulture) : "";
                            rating = res.VoteAverage.ToString(CultureInfo.CurrentCulture);
                            overview = res.Overview;
                            poster = res.PosterPath != null ? SmallPosterBase + res.PosterPath : null;
                            break;
                        }
                }
                Logger.Trace("Media Name: {MediaName}, Year: {Year}, Rating: {Rating}, Overview:{Pverview}, PosterPath: {Poster}",
                    mediaName, year, rating, overview, poster);
                items.Add(new ListItem(mediaName, year, rating, overview, poster, "", id.ToString(), mediaType));
                Logger.Info("Added {MediaName} to List", mediaName);
            }

        return items;
    }

        public static ObservableCollection<ListItem> ExtractTvDetailsIntoListItem(TvShow result)
        {
            var items = new ObservableCollection<ListItem>();
            var mediaName = result.Name;
                var year = result.FirstAirDate != null ? result.FirstAirDate.Value.Year.ToString(CultureInfo.InvariantCulture) : "";
                var rating = result.VoteAverage.ToString(CultureInfo.CurrentCulture);
                var overview = result.Overview;
                var poster = result.PosterPath != null ? SmallPosterBase + result.PosterPath : null;
                items.Add(new ListItem(mediaName, year, rating, overview, poster,"",result.Id.ToString()));

        return items;
    }

        public static ObservableCollection<ListItem> ExtractCollectionDetailsIntoListItem(
            Collection result)
        {
            var items = new ObservableCollection<ListItem>();
            var mediaName = result.Name;
                const string year = "";
                const string rating = "";
                const string overview = "";
                var poster = Convert.ToString(result.PosterPath != null ? SmallPosterBase + result.PosterPath : null, CultureInfo.InvariantCulture);
                items.Add(new ListItem(mediaName, year, rating, overview, poster, "", result.Id.ToString()));

        return items;
    }

        public static ObservableCollection<ListItem> ExtractMoviesDetailsIntoListItem(
            Movie result)
        {
            var items = new ObservableCollection<ListItem>();
            var mediaName = result.Title;
                var year = result.ReleaseDate != null ? result.ReleaseDate.Value.Year.ToString(CultureInfo.InvariantCulture) : "";
                var rating = result.VoteAverage.ToString(CultureInfo.CurrentCulture);
                var overview = result.Overview;
                var poster = result.PosterPath != null ? SmallPosterBase + result.PosterPath : null;
                items.Add(new ListItem(mediaName, year, rating, overview, poster, "", result.Id.ToString()));

        return items;
    }

        public static ObservableCollection<ListItem> ExtractTvDetailsIntoListItem(SearchContainer<SearchTv> result)
        {
            var items = new ObservableCollection<ListItem>();
            foreach (var item in result.Results)
            {
                var mediaName = item.Name;
                var year = item.FirstAirDate != null ? item.FirstAirDate.Value.Year.ToString(CultureInfo.InvariantCulture) : "";
                var rating = item.VoteAverage.ToString(CultureInfo.CurrentCulture);
                var overview = item.Overview;
                var poster = item.PosterPath != null ? SmallPosterBase + item.PosterPath : null;
                items.Add(new ListItem(mediaName, year, rating, overview, poster, "", item.Id.ToString()));
            }

        return items;
    }

    /// <summary>
    /// Prepares the Selected Result for Download And final List View
    /// </summary>
    /// <param name="result">Search Response</param>
    /// <param name="resultType">Type of search Response.</param>
    /// <param name="fullFolderPath">Full Path to the current Media Folder</param>
    /// <param name="rating">Rating for media</param>
    /// <param name="isPickedById"> identifies if Title was picked by media ID.</param>
    /// TODO: Merge parameter response and resultType.
    public void ResultPicked(dynamic result, string resultType, string fullFolderPath, string rating = "",bool isPickedById = false)
    {
        Logger.Debug("Preparing the Selected Result for Download And final List View");
        if (result.PosterPath == null)
        {
            Logger.Warn("No Poster Found, path - {Folder}", fullFolderPath );
            throw new InvalidDataException("NoPoster");
        }
        
        Logger.Debug("Rating: {Rating}, Result Type: {ResultType}", rating, resultType);
        var id = 0;
        var type = resultType;
        if (string.IsNullOrWhiteSpace(rating) && resultType != MediaTypes.Collection)
        {
            Logger.Debug("Rating is null or empty, getting rating from result");
            rating = result.VoteAverage.ToString(CultureInfo.InvariantCulture);
            Logger.Debug("Rating: {Rating}", rating);
        }

        var folderName = Path.GetFileName(fullFolderPath);
        var localPosterPath = fullFolderPath + @"\" + folderName + ".png";

        string posterUrl = string.Concat(PosterBase, result.PosterPath.Replace("http://image.tmdb.org/t/p/w500",""));

        if (resultType == MediaTypes.Tv)
        {
            dynamic pickedResult = isPickedById ? (TvShow)result : (SearchTv)result;
            var year = pickedResult.FirstAirDate != null ? pickedResult.FirstAirDate.Year.ToString(CultureInfo.InvariantCulture) : "";
            Util.AddToPickedListDataTable(_listDataTable, localPosterPath, pickedResult.Name,
                rating, fullFolderPath, folderName,
                year);
            id = pickedResult.Id;
        }
        else if (resultType == MediaTypes.Movie)
        {
            dynamic pickedResult = isPickedById ? (Movie)result : (SearchMovie)result;
            var year = pickedResult.ReleaseDate != null ? pickedResult.ReleaseDate.Year.ToString(CultureInfo.InvariantCulture) : "";
            Util.AddToPickedListDataTable(_listDataTable, localPosterPath, pickedResult.Title,
                rating, fullFolderPath, folderName, year);
            id = pickedResult.Id;
        }
        else if (resultType == MediaTypes.Collection)
        {
            dynamic pickedResult = isPickedById ? (Collection)result : (SearchCollection)result;
            Util.AddToPickedListDataTable(_listDataTable, localPosterPath, pickedResult.Name, rating, fullFolderPath,
                folderName, "");
            id = pickedResult.Id;
        }
        else if (resultType == MediaTypes.Mtv)
        {
            MediaType mediaType = result.MediaType;
            switch (mediaType)
            {
                case MediaType.Tv:
                {
                    type = "TV";
                    dynamic pickedResult = isPickedById ? (TvShow)result : (SearchTv)result;
                    var year = pickedResult.FirstAirDate != null
                        ? pickedResult.FirstAirDate.Year.ToString(CultureInfo.InvariantCulture)
                        : "";
                    Util.AddToPickedListDataTable(_listDataTable, localPosterPath, pickedResult.Name,
                        rating, fullFolderPath, folderName,
                        year);
                    id = pickedResult.Id;
                    break;
                }
                case MediaType.Movie:
                {
                    type = "Movie";
                    dynamic pickedResult = isPickedById ? (Movie)result : (SearchMovie)result;
                    var year = pickedResult.ReleaseDate != null
                        ? pickedResult.ReleaseDate.Year.ToString(CultureInfo.InvariantCulture)
                        : "";
                    Util.AddToPickedListDataTable(_listDataTable, localPosterPath, pickedResult.Title,
                        rating, fullFolderPath, folderName,
                        year);
                    id = pickedResult.Id;
                    break;
                }
            }
        }

        if (!isPickedById && id != 0)
        {
            FileUtils.SaveMediaInfo(id, type, fullFolderPath);
        }
        var tempImage = new ImageToDownload
        {
            LocalPath = localPosterPath,
            RemotePath = new Uri(posterUrl)
        };
        _imgDownloadList.Add(tempImage);
    }

    /// <summary>
    /// Searches TMDB for a query in Specified search mode
    /// </summary>
    /// <param name="query">Title to search</param>
    /// <param name="searchMode">Search Mode such as Movie,TV</param>
    /// <returns>Returns Search result with its Media Type</returns>
    public async Task<ResultResponse> SearchAsync(string query, string searchMode)
    {
        Logger.Info("Searching for {Query} in {SearchMode}", query, searchMode);
        object r = null;
        var mediaType = "";
        if (searchMode == MediaTypes.Movie)
        {
            if (query.ToLower(CultureInfo.InvariantCulture).Contains("collection"))
            {
                r = _serviceClient.SearchCollectionAsync(query).Result;
                mediaType = MediaTypes.Collection;
            }
            else
            {
                r = await _serviceClient.SearchMovieAsync(query);
                mediaType = MediaTypes.Movie;
            }
        }
        else if (searchMode == MediaTypes.Tv)
        {
            r = _serviceClient.SearchTvShowAsync(query).Result;
            mediaType = MediaTypes.Tv;
        }
        else if (searchMode == MediaTypes.Mtv)
        {
            r = _serviceClient.SearchMultiAsync(query).Result;
            mediaType = MediaTypes.Mtv;
        }

        var response = new ResultResponse
        {
            Result = r,
            MediaType = mediaType
        };
        return response;
    }

    /// <summary>
    /// Searches TMDB media by ID as per media Type.
    /// </summary>
    /// <returns>Returns Search result with its Media Type</returns>
    public ResultResponse SearchByIdAsync(int id, string mediaType)
    {
        Logger.Info("Searching for {Id} in {MediaType}", id, mediaType);
        object r = null;
        if (mediaType == MediaTypes.Movie)
        {
            r = _serviceClient.GetMovieAsync(id).Result;
        }
        else if (mediaType == MediaTypes.Collection)
        {
            r = _serviceClient.GetCollectionAsync(id).Result;
        }
        else if (mediaType == MediaTypes.Tv)
        {
            r = _serviceClient.GetTvShowAsync(id).Result;
        }
        var response = new ResultResponse
        {
            Result = r,
            MediaType = mediaType
        };
        return response;
    }
}